### PR TITLE
[NEW Model] Add RoFormerV2

### PR DESCRIPTION
**Pull request by JunnYu**
### PR types

New features

### PR changes

Models 

### Description

Add RoFormerV2

### Reference:
- keras: https://github.com/ZhuiyiTechnology/roformer-v2
- pytorch: https://github.com/JunnYu/RoFormer_pytorch

### Compare:
```python
import torch
import paddle
torch.set_grad_enabled(False)
paddle.set_grad_enabled(False)
paddle.set_device('cpu')
from transformers import BertTokenizer
from roformer import RoFormerModel
from paddlenlp.transformers import RoFormerv2Model
text = [
    "人生的路丰富而漫长,就像在宇宙中穿梭。",
    "时而太空漫步,无拘无束;时而速度失控,遭遇窘境。这意味着，漫漫人生路，同时涌现着异彩纷呈和艰难险阻。",
    "不知不觉已然忙碌了半生，对人生这道深刻的命题，也有了层次不一的理解。",
    "十八岁：成长成人，前方有路",
]
tokenizer = BertTokenizer.from_pretrained("roformer_v2_chinese_char_base")
pt_inputs = tokenizer(text, return_tensors="pt", padding=True, max_length=256)
pd_inputs = {k:paddle.to_tensor(v.numpy(), dtype="int64") for (k, v) in pt_inputs.items()}
pt_model = RoFormerModel.from_pretrained("roformer_v2_chinese_char_base")
pt_model.eval()
pd_model = RoFormerv2Model.from_pretrained("roformer_v2_chinese_char_base")
pd_model.eval()
pt_outputs = pt_model(**pt_inputs)[0].numpy()
pd_outputs = torch.tensor(pd_model(**pd_inputs).numpy())
print("paddle vs pytorch")
print("mean diff :", (pd_outputs - pt_outputs).abs().mean())
print("max diff :", (pd_outputs - pt_outputs).abs().max())
# paddle vs pytorch
# mean diff : tensor(7.8678e-07)
# max diff : tensor(1.6212e-05)
```
### CovertWeights
[模型配置文件.zip](https://github.com/PaddlePaddle/PaddleNLP/files/8620101/default.zip)
- download `模型配置文件.zip` and `unzip`
以`roformer_v2_chinese_char_base`为例：
```python
# 切换路径
cd roformer_v2_chinese_char_base
# 下载权重
wget https://huggingface.co/junnyu/roformer_v2_chinese_char_base/resolve/main/pytorch_model.bin
# 转换，即可得到paddle权重。
python convert.py
```

### CODE REVIEW

**Comment by JunnYu** [comment]
## CLUE 评测结果

使用多种**中文**预训练模型微调在 CLUE 分类任务下各验证集上有如下结果：

| Model                         | Config   | AVG       | AFQMC     | TNEWS     | IFLYTEK   | CMNLI     | OCNLI     | CLUEWSC2020 | CSL       |
| ----------------------------- | -------- | --------- | --------- | --------- | --------- | --------- | --------- | ----------- | --------- |
| RoBERTa-wwm-ext-large         | 24L1024H | 76.36     | 76.00     | 59.33     | 62.02     | **83.88** | 78.81     | **90.79**   | 83.67     |
| RoBERTa-wwm-ext               | 12L768H  | 74.46     | 74.60     | 58.08     | 61.23     | 81.11     | 76.92     | 88.49       | 80.77     |
| ERNIE 3.0-Base-zh             | 12L768H  | 75.38     | 75.93     | 58.26     | 61.56     | 83.02     | 80.10     | 86.18       | 82.63     |
| RBT6, Chinese                 | 6L768H   | 70.99     | 73.15     | 56.62     | 59.68     | 79.26     | 73.15     | 75.00       | 80.04     |
| ERNIE 3.0-Medium-zh           | 6L768H   | 72.78     | 73.37     | 57.00     | 60.67     | 80.64     | 76.88     | 79.28       | 81.60     |
| RoFormer-v2_chinese_char_base | 12L768H  | **76.87** | **77.06** | **59.53** | **63.76** | 81.58     | **81.56** | 87.83       | **86.73** |

注：其中**RoFormer-v2**使用了如下的分类头。

```python
class RoFormerv2ClassificationHead(Layer):
    def __init__(self, hidden_size, hidden_dropout_prob, num_classes):
        super().__init__()
        self.dense = nn.Linear(hidden_size, hidden_size)
        self.dropout = nn.Dropout(hidden_dropout_prob)
		self.classifier = nn.Linear(hidden_size, num_classes)
        
    def forward(self, features):
        x = features[:, 0]  # take [CLS] token
        x = self.dropout(x)
        x = self.dense(x)
        x = F.relu(x)
        x = self.dropout(x)
        x = self.classifier(x)
        return x
```
以上每个下游任务的固定超参配置如下表所示：

| TASK              | AFQMC | TNEWS | IFLYTEK | CMNLI | OCNLI | CLUEWSC2020 | CSL  |
| ----------------- | ----- | ----- | ------- | ----- | ----- | ----------- | ---- |
| epoch             | 3     | 3     | 3       | 2     | 5     | 50          | 5    |
| max_seq_length    | 128   | 128   | 128     | 128   | 128   | 128         | 128 |
| warmup_proportion | 0.1   | 0.1   | 0.1     | 0.1   | 0.1   | 0.1         | 0.1  |
| num_cards         | 1     | 1     | 1       | 1     | 1     | 1           | 1    |

不同预训练模型在下游任务上做 Grid Search 之后的最优超参（learning_rate、batch_size）如下：

| Model                         | AFQMC   | TNEWS   | IFLYTEK | CMNLI   | OCNLI   | CLUEWSC2020 | CSL     |
| ----------------------------- | ------- | ------- | ------- | ------- | ------- | ----------- | ------- |
| RoFormer-v2_chinese_char_base | 1e-5,64 | 1e-5,64 | 5e-5,32 | 3e-5, 64 | 1e-5,16 | 2e-5,16     | 1e-5,32 |

**Comment by JunnYu** [comment]
验证集高是理所当然，毕竟评估如此之密集，**100step**一次。。。。
但是线上测试集整体效果不太理想，输出的结果无法直接提交到clue排行榜，还需要修改文件名以区分1.0的版本还是1.1的版本。
image

比pytorch版本低了差不多2个点。（可能是pytorch训练的epoch多，基本上都为10个epoch，pytorch是**1个epoch评估一次**）
image

**Comment by ZeyuChen** [comment]
https://github.com/PaddlePaddle/PaddleNLP/tree/develop/examples/benchmark/clue
@JunnYu 可否在这个clue benchmark脚本上验证下你这个模型的指标？以及是否这个benchmark能快速产出一个可以在clue上提交通过的结果？

**Comment by JunnYu** [comment]
> ## CLUE 评测结果
> 使用多种**中文**预训练模型微调在 CLUE 分类任务下各验证集上有如下结果：
> 
> Model	Config	AVG	AFQMC	TNEWS	IFLYTEK	CMNLI	OCNLI	CLUEWSC2020	CSL
> RoBERTa-wwm-ext-large	24L1024H	76.36	76.00	59.33	62.02	**83.88**	78.81	**90.79**	83.67
> RoBERTa-wwm-ext	12L768H	74.46	74.60	58.08	61.23	81.11	76.92	88.49	80.77
> ERNIE 3.0-Base-zh	12L768H	75.38	75.93	58.26	61.56	83.02	80.10	86.18	82.63
> RBT6, Chinese	6L768H	70.99	73.15	56.62	59.68	79.26	73.15	75.00	80.04
> ERNIE 3.0-Medium-zh	6L768H	72.78	73.37	57.00	60.67	80.64	76.88	79.28	81.60
> RoFormer-v2_chinese_char_base	12L768H	**76.87**	**77.06**	**59.53**	**63.76**	81.58	**81.56**	87.83	**86.73**
> 注：其中**RoFormer-v2**使用了如下的分类头。
> 
> ```python
> class RoFormerv2ClassificationHead(Layer):
>     def __init__(self, hidden_size, hidden_dropout_prob, num_classes):
>         super().__init__()
>         self.dense = nn.Linear(hidden_size, hidden_size)
>         self.dropout = nn.Dropout(hidden_dropout_prob)
> 		self.classifier = nn.Linear(hidden_size, num_classes)
>         
>     def forward(self, features):
>         x = features[:, 0]  # take [CLS] token
>         x = self.dropout(x)
>         x = self.dense(x)
>         x = F.relu(x)
>         x = self.dropout(x)
>         x = self.classifier(x)
>         return x
> ```
> 
> 以上每个下游任务的固定超参配置如下表所示：
> 
> TASK	AFQMC	TNEWS	IFLYTEK	CMNLI	OCNLI	CLUEWSC2020	CSL
> epoch	3	3	3	2	5	50	5
> max_seq_length	128	128	128	128	128	128	128
> warmup_proportion	0.1	0.1	0.1	0.1	0.1	0.1	0.1
> num_cards	1	1	1	1	1	1	1
> 不同预训练模型在下游任务上做 Grid Search 之后的最优超参（learning_rate、batch_size）如下：
> 
> Model	AFQMC	TNEWS	IFLYTEK	CMNLI	OCNLI	CLUEWSC2020	CSL
> RoFormer-v2_chinese_char_base	1e-5,64	1e-5,64	5e-5,32	3e-5, 64	1e-5,16	2e-5,16	1e-5,32

@ZeyuChen 这个分类任务的结果就是使用这里的代码（同样也进行了grid search）https://github.com/PaddlePaddle/PaddleNLP/tree/develop/examples/benchmark/clue 跑出的。
目前还没跑阅读理解类型的任务。

**Comment by JunnYu** [comment]
# 支持静态导出，下面是静态导出加测试代码。
```python
import paddle
import numpy as np
from paddlenlp.transformers.roformerv2.modeling import (
    RoFormerv2ForMaskedLM,
    RoFormerv2ForSequenceClassification,
    RoFormerv2ForTokenClassification,
    RoFormerv2ForQuestionAnswering,
    RoFormerv2ForMultipleChoice,
)


class Predictor:
    def __init__(self, model_file, params_file, device="cpu"):
        config = paddle.inference.Config(model_file, params_file)
        if device == "gpu":
            config.enable_use_gpu(100, 0)
        elif device == "cpu":
            config.disable_gpu()

        config.switch_use_feed_fetch_ops(False)
        self.predictor = paddle.inference.create_predictor(config)
        self.input_handles = [
            self.predictor.get_input_handle(name)
            for name in self.predictor.get_input_names()
        ]
        self.output_handle = [
            self.predictor.get_output_handle(name)
            for name in self.predictor.get_output_names()
        ]

    def predict(self, input_ids):
        self.input_handles[0].copy_from_cpu(input_ids)
        self.predictor.run()
        outputs = [output_handle.copy_to_cpu() for output_handle in self.output_handle]
        return outputs[0] if len(outputs) == 1 else outputs


for model_cls in [
    RoFormerv2ForMaskedLM,
    RoFormerv2ForSequenceClassification,
    RoFormerv2ForTokenClassification,
    RoFormerv2ForQuestionAnswering,
    RoFormerv2ForMultipleChoice,
]:
    model = model_cls.from_pretrained("roformer_v2_chinese_char_small")
    model.eval()
    model = paddle.jit.to_static(
        model,
        input_spec=[
            paddle.static.InputSpec(shape=[None, None], dtype="int64"),  # input_ids
        ],
    )
    paddle.jit.save(model, "./tmp/infer_model")
    predictor = Predictor(
        "./tmp/infer_model.pdmodel", "./tmp/infer_model.pdiparams", "gpu"
    )
    input_ids = np.random.randint(1, 1234, size=(4, 32)).astype(np.int64)
    print(model_cls.__name__)
    output = predictor.predict(input_ids)
    if isinstance(output, list):
        for o in output:
            print(o.shape)
    else:
        print(output.shape)
    print("=" * 50)
# [2022-05-08 20:41:39,299] [    INFO] - Weights of RoFormerv2ForMaskedLM not initialized from pretrained model: ['cls.decoder_weight']
# RoFormerv2ForMaskedLM
# (4, 32, 12000)
# ==================================================
# [2022-05-08 20:41:52,546] [    INFO] - Weights of RoFormerv2ForSequenceClassification not initialized from pretrained model: ['classifier.weight']
# RoFormerv2ForSequenceClassification
# (4, 2)
# ==================================================
# [2022-05-08 20:42:06,546] [    INFO] - Weights of RoFormerv2ForTokenClassification not initialized from pretrained model: ['classifier.weight']
# RoFormerv2ForTokenClassification
# (4, 32, 2)
# ==================================================
# [2022-05-08 20:42:20,012] [    INFO] - Weights of RoFormerv2ForQuestionAnswering not initialized from pretrained model: ['classifier.weight']
# RoFormerv2ForQuestionAnswering
# (4, 32)
# (4, 32)
# ==================================================
# [2022-05-08 20:42:33,057] [    INFO] - Weights of RoFormerv2ForMultipleChoice not initialized from pretrained model: ['classifier.weight']
# RoFormerv2ForMultipleChoice
# (2, 2)
# ==================================================
```

**Comment by JunnYu** [comment]
| Model                 | Config   | CMRC2018        | CHID      | C<sup>3</sup> |
| --------------------- | ---------- | --------------- | --------- | ------------- |
| RoFormer-v2_chinese_char_base   | 12L768H  |  72.01/90.55 | 81.85 | 85.61    |

参数：
- crmr: 2e-5, 24
- c3: 1e-5, 32
- chid: 3e-5, 24 (没跑完，只跑了一个，数据集太大了)

**Comment by yingyibiao** [comment]
权重和词表均已上传。

File: paddlenlp/transformers/roformerv2/modeling.py [hunk 1]
```diff
@@ -0,0 +1,918 @@
...
+            attention_mask = paddle.unsqueeze(
+                (input_ids == self.pad_token_id
+                 ).astype(paddle.get_default_dtype()) * -1e4,
+                axis=[1, 2])
+        # For 2D attention_mask from tokenizer
+        elif attention_mask.ndim == 2:
+            attention_mask = paddle.unsqueeze(
+                attention_mask, axis=[1, 2]).astype(paddle.get_default_dtype())
+            attention_mask = (1.0 - attention_mask) * -1e4
+        attention_mask.stop_gradient = True
```
Comment (yingyibiao) [inline paddlenlp/transformers/roformerv2/modeling.py]: 这里需要加上 attention_mask 为 3D 的处理

**Review by yingyibiao — DISMISSED** [review]

File: paddlenlp/transformers/roformerv2/modeling.py (hunk 1 above)
Comment (JunnYu) [reply paddlenlp/transformers/roformerv2/modeling.py]: 目前只支持2D的attention mask。

Comment (yingyibiao) [reply paddlenlp/transformers/roformerv2/modeling.py]: 后续我们需要支持。

**Review by yingyibiao — APPROVED** [review]
LGTM
